### ISSUE TITLE
Make FHIR Data available to R

### ISSUE THREAD

**Issue by grahamegrieve** [ropensci/ozunconf17#3]
FHIR is a data format/RESTful API specification that is expected to be the main future method for exchange of data in the operational healthcare space. I'm interested in an R package that provides direct access to FHIR data so that people can easily use R for analysis of healthcare data without having to do any/much work to massage the data into a processible format. 

FHIR spec: http://hl7.org/fhir/
Proposal for big data access to health record systems: http://www.healthintersections.com.au/?p=2689

**Comment by njtierney** [comment]
Wow, this looks really great. Do you happen to have any example raw output data from FHIR handy?

**Comment by grahamegrieve** [comment]
well, there's small examples - nothing really big. But here's one example:
http://hl7.org/fhir/patient-operations.html#8.1.18.2.1. Or there is the ndjson format attached. I'll have better data by the day

[ndjson.zip](https://github.com/ropensci/ozunconf17/files/1362136/ndjson.zip)

**Comment by njtierney** [comment]
OK sure thing - it sounds like this data is really rich and is used for some important tasks, but I'm not really sure what they are. Do you think you could tell us a little bit more about who uses this sort of data, how it is measured, and what sort of analysis it might be used for?

**Comment by grahamegrieve** [comment]
FHIR is (becoming) the primary exchange format for operational healthcare data - that is, your actual personal healthcare information. It's the data that supports actual healthcare  so it's used by doctors, nurses, patients, carers. It includes things like patient identity, encounter details, patient problem list, medication details (prescribed, given), procedures, referrals, lab tests, imaging reports, care plans, etc. if you want to know what really happened, this is the data you want (unlike trial data which is often sanitised data .e.g. one arbitrarily value for blood pressure at admission, instead of the actual series of blood pressures measured that change constantly as you comfort/panic levels change, your posture changes, you go on a drip, etc - they tell a real story). This data you use for assessing care outcomes, care efficiency. Merged with claims / billing data it allows you to compare complexity with cost, and it's also the data that allows you to evaluate turn arounds, connect to outcomes, etc. 

Most of the data that is currently available about healthcare is aggregate data that glosses over the fine details, and the analysis is only as good as the extraction assumptions.  FHIR is a format that is used for exchange between the operational systems but more and more people are getting interested in accessing the data for analysis. Hence my interest here: can we make it dead easy to get at the data from R? (actually, other people tell me this would be good, I just want to make them happy)

**Comment by grahamegrieve** [comment]
Yes absolutely that would be a great demo - saw a company working on that with properatrary tools this week

**Comment by adam-gruer** [comment]
Perhaps the first step is building a parser in R for this data. In my head this would begin with the application of well established R packages to the FHIR domain to produce tidy data from the source xml and/or JSON.
Wrap it all up in a cool sounding package like tidyFHIR (ok, cooler sounding than that) and we'd look like wizards*
* I am prone to oversimplifying/underestimating

**Comment by grahamegrieve** [comment]
Figuring out what the steps are, and writing accessors for the FHIR data is what I hope to do at this Unconference

**Comment by adam-gruer** [comment]
@grahamegrieve  I work at the Royal Children's Hospital (RCH).  We have implemented the Epic electronic medical record. I'm almost afraid to ask but does Epic support the FHIR standard?  Let me know if you have any contact at RCH or the Murdoch Children's Research Institute (MCRI) and I'll try and do a little back grounding with them before next week.  🏥 💻 📦 🎊 

![data](https://media.giphy.com/media/3o7TKSx0g7RqRniGFG/giphy.gif)

**Comment by grahamegrieve** [comment]
Epic have implemented FHIR, and are tracking along this path, though they haven't yet announced when they might make it available to RCH. I have no contacts at RCH presently.

**Comment by robmulders** [comment]
Nice to see there is interest for getting data into R! We have been working on it since the HL7 WGM in May 2017. After a few attempts to link in the DLL from the Furore .Net API into R, we decided to step back to native R, to reduce complexity.

Our first 0.1 version has been released by my colleague Sander Laverman last week to CRAN, the standard place for R packages. The package is called R on FHIR and supports read and search, including some supporting functions. You can find it here: https://cran.r-project.org/web/packages/RonFHIR/index.html

Sander is a new member of the Furore FHIR team. He has been working extensively with R during his study Medical Informatics in Amsterdam. Sander is also the author of SNOLYZE, a SNOMED CT Expression Constraint Language Execution Engine in R.

We aim to support researchers in healthcare with our R on FHIR package. All feedback is welcome! I will discuss with the Furore FHIR team this week which channel we will use for discussions, support and feedback.

**Comment by adam-gruer** [comment]
wow such ⌚️ !
One mini project we could consider at ozunconf is creating a vignette for Sander's package.

**Comment by rienwertheim** [comment]
Let's explore if we can do something at FHIR DevDays with this as well.

**Comment by slaverman** [comment]
@adam-gruer, a vignette would be a great idea!

For version 0.2 of R on FHIR we are looking for possible ways to help the user extract the desired data from the resulting 'tidy data.frames', see picture below how a bundle is presented in R. 

![tidy dataframe](https://user-images.githubusercontent.com/19187321/31888070-e1e852d8-b7fa-11e7-8e6b-6fc86ea88375.JPG)

We had the following options in mind:
- Object models for all StructureDefinitions (like our .NET API), see Object Model in the first picture below
- Functions to extract the desired data from the resulting data.frames, see Functional in the first picture below
- Function to flatten the whole data.frame, see second picture below
- Leave it as it is and let the user decide on how they extract data from the resulting data.frame

**Any feedback and other suggestions are welcome!**



![object vs functional](https://user-images.githubusercontent.com/19187321/31888356-f95113f0-b7fb-11e7-9546-7d8c8e34fa81.JPG)

![flatten dataframe](https://user-images.githubusercontent.com/19187321/31888147-2f15d9f4-b7fb-11e7-9d25-99ce85fa37cb.JPG)

**Comment by adam-gruer** [comment]
@grahamegrieve , this is a useful list of existing r 📦 for working with json and api 
https://github.com/ropenscilabs/exploRingJSON/blob/master/README.md

**Comment by grahamegrieve** [comment]
we decided that the best place to apply ourselves is at the data acquisition level. Generating object models would really mainly help with dates and the input from R experts was that dates weren't a big deal where as   wrangling the data really was important. So I invested in the graphQL side of things to get the data normalised. But there would certainly be space for exemplar R functions for post-processing to do things you can't do in graphQL

**Comment by stefaniebutland** [comment]
@grahamegrieve's blog post on this: FHIR and R: ronfhir http://www.healthintersections.com.au/?p=2740
